STYLE.md

```diff
@@ -0,0 +1,4068 @@
+
+### Functional Options
+
+Functional options is a pattern in which you declare an opaque `Option` type
+that records information in some internal struct. You accept a variadic number
+of these options and act upon the full information recorded by the options on
+the internal struct.
+
+Use this pattern for optional arguments in constructors and other public APIs
+that you foresee needing to expand, especially if you already have three or
+more arguments on those functions.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+// package db
+
+func Open(
+  addr string,
+  cache bool,
+  logger *zap.Logger
+) (*Connection, error) {
+  // ...
+}
+```
+
+</td><td>
+
+```go
+// package db
+
+type Option interface {
+  // ...
+}
+
+func WithCache(c bool) Option {
+  // ...
+}
+
+func WithLogger(log *zap.Logger) Option {
+  // ...
+}
+
+// Open creates a connection.
+func Open(
+  addr string,
+  opts ...Option,
+) (*Connection, error) {
+  // ...
+}
+```
+
+</td></tr>
+<tr><td>
+
+The cache and logger parameters must always be provided, even if the user
+wants to use the default.
+
+```go
+db.Open(addr, db.DefaultCache, zap.NewNop())
+db.Open(addr, db.DefaultCache, log)
+db.Open(addr, false /* cache */, zap.NewNop())
+db.Open(addr, false /* cache */, log)
+```
+
+</td><td>
+
+Options are provided only if needed.
+
+```go
+db.Open(addr)
+db.Open(addr, db.WithLogger(log))
+db.Open(addr, db.WithCache(false))
+db.Open(
+  addr,
+  db.WithCache(false),
+  db.WithLogger(log),
+)
+```
+
+</td></tr>
+</tbody></table>
+
+Our suggested way of implementing this pattern is with an `Option` interface
+that holds an unexported method, recording options on an unexported `options`
+struct.
+
+```go
+type options struct {
+  cache  bool
+  logger *zap.Logger
+}
+
+type Option interface {
+  apply(*options)
+}
+
+type cacheOption bool
+
+func (c cacheOption) apply(opts *options) {
+  opts.cache = bool(c)
+}
+
+func WithCache(c bool) Option {
+  return cacheOption(c)
+}
+
+type loggerOption struct {
+  Log *zap.Logger
+}
+
+func (l loggerOption) apply(opts *options) {
+  opts.logger = l.Log
+}
+
+func WithLogger(log *zap.Logger) Option {
+  return loggerOption{Log: log}
+}
+
+// Open creates a connection.
+func Open(
+  addr string,
+  opts ...Option,
+) (*Connection, error) {
+  options := options{
+    cache:  defaultCache,
+    logger: zap.NewNop(),
+  }
+
+  for _, o := range opts {
+    o.apply(&options)
+  }
+
+  // ...
+}
+```
+
+Note that there's a method of implementing this pattern with closures but we
+believe that the pattern above provides more flexibility for authors and is
+easier to debug and test for users. In particular, it allows options to be
+compared against each other in tests and mocks, versus closures where this is
+impossible. Further, it lets options implement other interfaces, including
+`fmt.Stringer` which allows for user-readable string representations of the
+options.
```

Comment:
As it was mentioned by @acud elsewhere, overusing options as we do it now may not always be good, so maybe some discussion needed regarding the alternatives, and some examples showing when to use options and when to use a simple approach.
Another problem is that the approach to options which is described here differs from what's in use in the most of go-spacemesh codebase at this moment, which is using closure-based options. Although the advantages described in this paragraph may be valid, they may not be always worth the cost of adding verbosity to already verbose options mechanism (which is essentially a verbose kwargs replacement), and also in this case we will either have inconsistent option mechanisms in the codebase, or we will have to refactor a lot of existing code.

Reply:
The difference between "closure options" and "functional options" is only that the closure options manipulate the created object directly, while the functional operate on an intermediary object that only holds configuration and than use that to construct the actual service. Sure in the details it is different but from the point of using them I see little difference.

Functional options should be used for _optional_ parameters and thats where they are currently overused in our codebase. In many places the parameters that are passed to the services are not optional and not setting them will cause the instantiation of the object to fail.

I've occasionally done this as well primarily for readability:

```go
func NewService(
    db,
    lg,
    clk,
    sync,
    h,
)
```

is just much less readable than

```go
func NewService(
    WithDB(db),
    WithLogger(lg),
    WithClock(clk),
    WithSyncer(sync),
    WithHandler(h),
)
```

Sure variable naming can be improved too to mitigate this issue but functional options have an additional benefit for testing where constructing many services with slightly different configuration becomes much easier:

```go
defaultOpts := []Opts{
    WithDB(db),
    WithLogger(lg),
    WithClock(clk),
    WithSyncer(sync),
    WithHandler(h),
}

testservice1 := NewService(defaultOpts..., WithHandler(mockHandler))
testservice2 := NewService(defaultOpts..., WithSyncer(mockSyncer))
// ... etc.
```

Reply:
Problem is that the paragraph I commented recommends against using closure-based options. Thus we either update the guide, refactor our code or accept it as an inconsistency

As of options being useful in many cases, I can only agree with that, but there are also arguments against using them too much:
https://github.com/spacemeshos/go-spacemesh/pull/6085#discussion_r1660225344

I for one don't find these arguments to be applicable to all cases as having 10 different constructors to avoid the use of the option pattern doesn't sound too good to me, but there probably needs to be some kind of balance

Reply:
So I'm with @ivan4th on this. Personally I find this sort of pattern to be very confusing because it is also allowing for irregular sequence of setting things in a constructor. It has the sort of flexibility that the language intentionally tries to avoid for various reasons.
Also having these sort of parameters has the unfortunate consequence that a lot of the components in the code-base do things in a very complex way. For example:
- component defines it's _own_ configuration with a default configuration which is then _potentially_ overridden by the instantiator (BAD) - configuration should never be an optional parameter
- logger is defined by the component but then _potentially_ overridden by the instantiator - why should a logger parameter be optional? for me this is a mandatory thing and that instantiator should know whether the logger that should be passed is a noop logger or not
- DB is not optional and should not be marked as such by using this pattern.

IMO the "internal vocalisation" of how we read code shouldn't affect the way we write it. And yes I agree that sometimes having a constructor call with a lot of arguments can become cumbersome but you at least know for sure what is needed and it usually also breaks-fast so one can fix it fast.

So TL;DR my personal opinion is that I find this pattern less readable and that it adds an unreasonable amount of mental taxation reading the code (and resolving its dependencies)